sources/VT100Grid.m

```diff
@@ -388,12 +388,12 @@ - (int)scrollUpIntoLineBuffer:(LineBuffer *)lineBuffer
 
 - (int)resetWithLineBuffer:(LineBuffer *)lineBuffer
         unlimitedScrollback:(BOOL)unlimitedScrollback
-        preserveCursorLine:(BOOL)preserveCursorLine {
+        preservePromptLines:(int)preservePromptLines {
     self.scrollRegionRows = VT100GridRangeMake(0, size_.height);
     self.scrollRegionCols = VT100GridRangeMake(0, size_.width);
     int numLinesToScroll;
-    if (preserveCursorLine) {
-        numLinesToScroll = cursor_.y;
+    if (preservePromptLines > 0) {
+        numLinesToScroll = cursor_.y - preservePromptLines + 1;
```

Comment:
Must ensure this does not go negative. `MAX(0, cursor_.y - preservePromptLines + 1)`.